internal/onboarding/exec.go

```diff
@@ -0,0 +1,480 @@
+	"github.com/rs/zerolog/log"
+	"go.mondoo.com/cnquery/v11/cli/components"
+	"go.mondoo.com/cnquery/v11/cli/theme"
+)
+
+// WriteHCL Writes the generated HCL code
+func WriteHCL(hcl string, location string, identifier string) (string, error) {
+	// Determine write location
+	dirname, err := getOutputDirPath(location, identifier)
+	if err != nil {
+		return "", err
+	}
+
+	// check if output location exists and if it's a file
+	outputDirLocation, err := os.Stat(dirname)
+	if !os.IsNotExist(err) && !outputDirLocation.IsDir() {
+		return "", fmt.Errorf("output location %s already exists and is a file", dirname)
+	}
+
+	// Create directory, if needed
+	if os.IsNotExist(err) {
+		directory := filepath.FromSlash(dirname)
+		if _, err := os.Stat(directory); os.IsNotExist(err) {
+			err = os.MkdirAll(directory, 0700)
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+
+	// Create HCL file
+	outputLocation := filepath.FromSlash(fmt.Sprintf("%s/main.tf", dirname))
+	err = os.WriteFile(filepath.FromSlash(outputLocation), []byte(hcl), 0700)
+	if err != nil {
+		return "", err
+	}
+
+	return dirname, nil
+}
+
+// getOutputDirPath determine output directory location based on how the output location was set
+func getOutputDirPath(location string, identifier string) (string, error) {
+	// determine code output path
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	// If location was passed, return that location
+	if location != "" {
+		return filepath.FromSlash(location), nil
+	}
+
+	// If location was not passed, assemble it with mondoo from os homedir
+	// @afiune we should have this location somewhere, look it up
+	return filepath.FromSlash(fmt.Sprintf(
+		"%s/%s/%s/%s/%s",
+		home, ".config", "mondoo", "onboarding", identifier,
+	)), nil
+}
+
+// Execute a terraform plan & apply
+func TerraformPlanAndExecute(workingDir string) error {
+	// Ensure Terraform is installed
+	tf, err := LocateOrInstallTerraform(false, workingDir)
+	if err != nil {
+		return err
+	}
+
+	vw := terminal.NewVerboseWriter(10)
+	tf.SetStdout(vw)
+	tf.SetStderr(vw)
+
+	// Initialize tf project
+	if err := TerraformInit(tf); err != nil {
+		return err
+	}
+	vw.Close()
+
+	vw = terminal.NewVerboseWriter(10)
+	tf.SetStdout(vw)
+	tf.SetStderr(vw)
+
+	// Write plan
+	changes, err := TerraformExecPlan(tf)
+	if err != nil {
+		return err
+	}
+	vw.Close()
+
+	// Display changes and determine if apply should proceed
+	proceed, err := DisplayTerraformPlanChanges(tf, *changes)
+	if err != nil {
+		return err
+	}
+
+	// If not proceed; display guidance on how to continue outside of this session
+	if !proceed {
+		fmt.Println(provideGuidanceAfterExit(true, tf.WorkingDir(), tf.ExecPath()))
+		return nil
+	}
+
+	vw = terminal.NewVerboseWriter(10)
+	tf.SetStdout(vw)
+	tf.SetStderr(vw)
+
+	// Apply plan
+	if err := TerraformExecApply(tf); err != nil {
+		return errors.New(provideGuidanceAfterFailure(err, tf.WorkingDir(), tf.ExecPath()))
+	}
+	vw.Close()
```

Comment:
is it an issue if these are not closed on error?

Reply:
Good question.

There is no issue, but I think it is nice to clear the terminal since we are extracting the error and presenting it to the user, this way it won't be duplicate.